### ISSUE TITLE
WebGL overlay?

### ISSUE THREAD

**Issue by jetic83** [openseadragon/openseadragon#1008]
There are several plugins for SVG and Canvas overlays (fabricjs, paper, canvas, svg...).

I experience significant slowness when using them with a lot of objects (e.g. 100k rectangles in my deep zoom image). However, WebGL can render 100k rectangles easily and fast (e.g. with three.js).

Are there also plugins for WebGL overlays? Then I could use that to overlay to add the rectangles.

Thanks!

**Comment by iangilman** [comment]
It seems like it should definitely be possible... I think just no one has done it. Whoever wants to try, I'm guessing starting with https://github.com/altert/OpenSeadragonCanvasOverlay would probably be easiest. Maybe do an OpenSeadragonThreeJsOverlay or something...

@jetic83 If you're not up for tackling it yourself, you might ask around in https://gitter.im/openseadragon/openseadragon to see if anyone is interested.

**Comment by thejohnhoffer** [comment]
Hi there- I've just developed a small plugin that filters openSeadragon tiles via WebGL. It still ultimately renders through the openSeadragon drawer, but it does let you use GLSL shaders via WebGL for processing the images (or generating rectangles).

Check it out here: [viaWebGL + openSeadragonGL](http://via.hoff.in)

A couple thoughts:
* It renders each tile to the same `WebGLRenderingContext` offscreen
* Then it plugs the rendered tile into the drawn Canvas via
    * `CanvasRenderingContext2D.drawImage(WebGLRenderingContext.canvas)`
        * On the `'tile-drawing'` event of `OpenSeadragon.Viewer`
    * `HTMLImageElement.src = WebGLRenderingContext.canvas.toDataURL()`
        * On the `'tile-loaded'` event of `OpenSeadragon.Viewer`

Since the steps in`'tile-drawing'` work faster than `'tile-loaded'`, I prefer to use `'tile-drawing'` even when I need to use GLSL shaders just once for each tile. I've set the default handler on `'tile-drawing'` to run only the first time it's called.

**Comment by iangilman** [comment]
Sounds like some interesting possibilities to explore!

I assume given the architecture of your plugin that it's only possible to modify the tiles, but not possible to overlay original geometry on top of the OSD world, is that correct? I guess if you had a webGL overlay that did zooming and panning and drew its own images directly to the screen that would handle that case. Of course #68 is another thing, but possibly related.

**Comment by thejohnhoffer** [comment]
Yes- my plugin just acts on the tiles. If you want to overlay original geometry, I suggested that you could do that with a second (empty) tileSource layer. It could even be a simple black image, or a repeated black tile. 

Looking at it now, you would want to send the `Tile.position` as a `uniform vec2` to GLSL shaders for each tile that you render. My plugin has support for sending custom variables like that into your shaders using webGL's native API. In the shaders, you would then just add `Tile.position` to your internal `v_tile_pos` to get the whole overlay to work as one continuous space.

**Comment by iangilman** [comment]
Right on... interesting stuff to explore!

**Comment by slntRohit** [comment]
Hi,
I know it's been two years but I have created a WebGL Threejs overlay for OpenSeaDragon. Demo [here](https://fervent-snyder-c66537.netlify.com/)
Repo [OpenSeadragonThreejsOverlay](https://github.com/slntRohit/OpenSeadragonThreejsOverlay)
@jetic83 @iangilman Does it satisfy the requirements of this Issue?

**Comment by jetic83** [comment]
Thank you very much @slntRohit, this looks very nice! Here are some thoughts:

- The disadvantage of e.g. the fabricjs plugin was that is becomes very slow if is has to render e.g. 5000+ different objects on the image (e.g. annotating cell nuclei on a pathology image). Then, navigation with OSD becomes slow and uncomfortable. Can you render more 5000+ objects on the image in a small region?

- Does your overlay rotate with image rotation OSD?

- Does your overlay flip with image flipping in OSD?

If so, this might be a very valuable plugin, indeed!

**Comment by jetic83** [comment]
Nice!

Flipping is the latest feature of OSD
https://github.com/openseadragon/openseadragon/pull/1441

Of course also combined with rotation

**Comment by jetic83** [comment]
Yes, @iangilman , when those basic controls (rotation, flipping) are supported, I strongly recommend to include this in the plugin list. The performance difference to e.g. fabricjs with a mass of objects is significant.

**Comment by iangilman** [comment]
@jetic83 Awesome; thank you for checking it out! 

@slntRohit I'd say go ahead and add it to the plugin list already, since it's already useful to people who don't need rotation and flipping support (though of course support for those features would be great!). Please make a pull request on: 

https://github.com/openseadragon/site-build/blob/0d63cdab413a0be9832c75fbce14f475c6386e95/www/index.html#L151

**Comment by Aiosa** [comment]
Is it possible to preserve transparency with webGL canvas? When registered on `tile-drawing` event, the areas set to `vec4(0, 0, 0, 0)` became black rather than show the content beneath. My question is - is there a way to prevent something in OpenSeadragon that clears the canvas? Or do I misunderstand something? Either way, with `tile-loaded` there is no such issue, though there are other loading-related artefact issues so I would prefer to use `tile-drawing` event.

**Comment by iangilman** [comment]
@Aiosa are your tiles in PNG format with transparency? That's the first step. 

I see you've also filed https://github.com/thejohnhoffer/viaWebGL/issues/10, which is good, since it sounds like the issue may be specific to that plugin.

**Comment by Aiosa** [comment]
@iangilman  Partially I was able to solve it, though I don't know if it does not create some overhead. Plugin passes the data with `tile-loaded` event in URL which was pretty slow and also created visual artifacts on loading. I am still unsure how OSD works internally - for each tile source, there is one layer of canvas that is drawn onto? Or do all the tilesources share one layer?

1) on `tile-loaded` event, this code works, but I was (and am) concerned about the canvas creations...
```
output = drawGL(...)  // canvas object

var canvas = document.createElement( 'canvas' )
canvas.width = e.tile.sourceBounds.width;
canvas.height = e.tile.sourceBounds.height;
var renderedContext = canvas.getContext('2d');
renderedContext.drawImage(output, 0, 0);
e.tile.context2D = renderedContext;  
```

2) whereas this (I tried to re-use objects I found, though I was not sure I am doing the right thing but I did not like creating new canvas for each tile - I thought that OSD already prepared canvas for the source to be drawn onto and I am creating another one, but that did not work. I tried to use the webGL canvas too, the transparency issue remained
```
var renderedContext = output.getContext('2d');
e.tile.context2D = renderedContext;
```

3) and in `tile-drawing` event also remained opaque:
```
//e - event argument
 e.rendered.drawImage(output...);
```

Not sure why.

My tiles come from pyramidal TIFs -> on-the-fly generated DZI via IIPImage server. The data comes as RGBA. But even if it didn't,  webGL will make it transparent anyway..? I mean the tiled image is used only as a texture input to webGL, so... I would be very surprised if it was not in a transparency-supporting format.


Furthermore, when using a solution from 1) the code seems to be having problems with refresh request - when a user changes threshold value and I need to redraw everything, I call this (I considered using cached tiles too, but not sure how to do that yet):
```
 viewer.addTiledImage({
          tileSource : layerTileSource,
          index: index,  
          replace: true
          });
```
and it works fine. The **very same** code for different control value produces many lines of

> Ignoring tile h.Tile loaded before reset: URL

and in debug mode, the color of tiles changes. Also when playing with the second value for a while, the viewer becomes slower, it seems new tiles were stacked atop and not replaced. That brings me back to the question of how optimal is the 1.st approach.

PS: I am very concerned with performance, that's why I went with WebGL.

**Comment by iangilman** [comment]
My understanding of https://github.com/thejohnhoffer/viaWebGL is that it doesn't convert the entire drawing process to WebGL; it just allows you to use WebGL shaders to modify the tiles. @thejohnhoffer please correct me if I'm wrong. So I don't expect that that plugin would improve performance. @Aiosa are you using the shader modification facility of that plugin?

I think the first step, if you want transparency, is to get things working with transparency without the plugin, to remove that factor for starters. You say the file is a TIFF being served via IIPImage... do you know what format IIPImage is turning the tiles into? If it's using JPG, then you're losing the alpha data. Can you configure it to send them via PNG? TIFF isn't a web format, so some translation is needed.

**Comment by Aiosa** [comment]
Yeah, it does not convert the process of drawing to webGL. However, I am not drawing the tile data I get, I am processing it (three different values stored in RGB) and it can be faster with shaders rather than plain javascript (per-pixel tile processing, edge detection etc...). I am not sure what format is being sent to me, however, I **know** the incoming data has four components: RGBA. The transparency works without a plugin - it already worked, I am not the official author of that project I am just improving performance and user experience (original project here: https://github.com/cesekova/openseadragon_project).

**Comment by Aiosa** [comment]
Maybe it would be cool to also consider replacing hard-coded target for tile (`canvas` or `div`) with a custom class provider that would allow plugins to override the behaviour of drawing a tile. It could just specify the interface and one method `registerCustomProvider(...)` or something similar that would behave as needed.

e.g.
```
    /**
     * A &lt;canvas&gt; element if the browser supports them, otherwise a &lt;div&gt; element.
     * Child element of {@link OpenSeadragon.Drawer#container}.
     * @member {Element} canvas
     * @memberof OpenSeadragon.Drawer#
     */
    this.canvas     = $.makeNeutralElement( this.useCanvas ? "canvas" : "div" );
```
could become
```
    this.canvas     = provider.makeNeutralElement();
```
or even do not specify the canvas at all and use `provider.canvas` instead... I thing is I considered modifying OSD for better performance to use WebGL directly, but this would prevent me from using future OSD updates...

**Comment by iangilman** [comment]
Yes, having a WebGL renderer would be awesome! There's discussion about that in #68 and #1482, but as far as I know there hasn't been any progress on making it happen. I expect it'll be a good chunk of work. If you're interested in pursuing it, I'm happy to help you understand the code. 

And yes, it would be great if we had a better mechanism for creating/using tile data on the fly. Some discussion about that in #119 and #1707. I believe it's possible at the moment, but not very smooth. I do think you're close with modifying `e.rendered`, though as you can see here: 

https://github.com/openseadragon/openseadragon/blob/a6792138814d1eff5ae4fca526b989fc917245be/src/tile.js#L378

...you're just modifying the event object, which isn't used again. To make it possible to change to a different context, we would need to update that code a little somehow (like grabbing the new value out of the event after it's been called). You could certainly do a pull request with that change... it's pretty targeted. 

That said, the `drawingHandler` mechanism happens every frame... is that what you want? If not, you could modify the tile's `context2D` instead, which would carry through from frame to frame.

**Comment by BeebBenjamin** [comment]
Hello!  I am working on a threejs overlay plugin for Mirador as part of a work project at the Bodleian Library to incorporate 2.5D data into Digitial Bodleian.  Most of my time at the moment is researching how OpenSeadragon works, hence why I am asking this question here.  

Due to the limitations of the size the WebGL renderer can be in Chrome and Firefox (circa max 4k without a performance hit or the renderer truncating the content), I decided to map the renderer dimensions I have in my overlay to the calculated intersection of the image present on screen and this works great.  However, the spring animation kicks in on pan and zoom which means my intersection calculations are always from the beginning of the animation, and this creates a lag until the animation finishes.

I am using a 'viewport-change' handler to get the calculations before updating the threejs overlay and this seems to create the smoothest response in updating the renderer dimensions in comparison to the other OSD handlers I have tried.  However, I am wondering if another handler may work better?  I was also wondering if anyone knows of a way to get the future position and intersection for the final frame of an animation sequence, or if this is even possible?  Otherwise, I will impement a clever way to mask the problem.  I've attached a small image during panning, the darker area is the renderer rendering surface normals, and the lighter image is the OSD image beneath the overlay.

![image](https://user-images.githubusercontent.com/4963887/179797540-0126d423-af57-457b-8e7e-da5370d41ef1.png)

**Comment by Aiosa** [comment]
Hi, as long as you can have the overlay tiled, e.g. treated as a TileSource/TiledImage,  so far the best solution I know of  (including available code for OSD I found) is in my small WebGL module within a Pathology Viewer. It uses several ways of binding WebGL to the TiledImage instance, including the new [data pipeline](https://github.com/openseadragon/site-build/pull/258) that [will get released](https://github.com/openseadragon/openseadragon/pull/2148) in the next OSD version. Unlike all solutions you can find it also supports caching (with the new API).

Follow https://github.com/RationAI/pathopus/tree/development/modules/webgl this link to the module. The `[webGLToOSDBridge.js](https://github.com/RationAI/pathopus/blob/development/modules/webgl/webGLToOSDBridge.js)` file contains logics of binding the WebGL rendering to the `TiledImage` instance. Maybe you will find the WebGL library useful as well :)

**Comment by iangilman** [comment]
@BeebBenjamin it sounds like you are basing your positioning off of the `viewer.viewport.getBounds()` which gives you the target bounds for the animation, not the current bounds. To get the exact current bounds for this frame, use `viewer.viewport.getBounds(true)`.

**Comment by iangilman** [comment]
BTW, for anyone interested in WebGL overlays, you might like to know that @rsimon is adding them to https://recogito.github.io/annotorious/: 

https://twitter.com/aboutgeo/status/1547896108273152000

**Comment by rsimon** [comment]
Hi all,

thanks @iangilman for pinging me. Yes, I'm working on a WebGL overlay plugin (2D only though, not 2.5D!) It's running pretty well by now. It handles rotation, but I still need to look into flip. (Shouldn't be a big change though.) Here's my latest test with 11.5k overlays (a mix of polygons and rectangles). I haven't tested out the limitations yet, but it looks like you can safely draw  significantly more shapes than that. (I did make a simple test with one million rectangles when I started. Seemed to work ok.)

I'm using [PixiJS](https://pixijs.com/) for rendering, BTW, rather than WebGL directly, which seems to abstract away many of the low-level details. (You're essentially dealing with a simple scene graph.) In my case, I'm attaching the render call to the OSD `update-viewport` event.

**Comment by Aiosa** [comment]
@rsimon wow looks awesome. Maybe I will try your annotations as well :)

**Comment by BeebBenjamin** [comment]
> @BeebBenjamin it sounds like you are basing your positioning off of the `viewer.viewport.getBounds()` which gives you the target bounds for the animation, not the current bounds. To get the exact current bounds for this frame, use `viewer.viewport.getBounds(true)`.

Thanks so much, this has fixed this problem for me!

**Comment by BeebBenjamin** [comment]
I had a hunch, and `viewer.viewport.getZoom` was the culprit.  This is now set to `true` and gets the current zoom!  Many thanks for your replies.  The project at the Bodleian is called ARCHiOx and as part of it we are researching WebGL either directly or abstracted via things like Threejs and PixiJS to render normal maps with interactive lighting.  Hopefully I will be able to share the output of our research with the community soon!

**Comment by iangilman** [comment]
@rsimon 

> But I'm planning to make things more modular this time, and hoping to release the overlay drawing feature as a separate OSD plugin that can be used without Annotorious, too.

That would be wonderful!!

**Comment by iangilman** [comment]
@BeebBenjamin Great! Glad that sorted it out. ARCHiOx looks like a fantastic project! Please do share it when there's something to see... I'd love to see it and I'm sure others in the OSD community would as well :)

**Comment by iangilman** [comment]
Might be interesting to revisit this now that #2310 has landed.
